portal-bridge/src/census.rs

```diff
@@ -0,0 +1,287 @@
+use alloy_primitives::U256;
+use anyhow::anyhow;
+use delay_map::HashMapDelay;
+use discv5::enr::NodeId;
+use futures::{channel::oneshot, StreamExt};
+use tokio::{
+    sync::mpsc,
+    time::{Duration, Instant},
+};
+use tracing::{error, info};
+
+use ethportal_api::{
+    generate_random_remote_enr,
+    jsonrpsee::http_client::HttpClient,
+    types::{
+        distance::{Distance, Metric, XorMetric},
+        portal::PongInfo,
+    },
+    BeaconContentKey, BeaconNetworkApiClient, Enr, HistoryContentKey, HistoryNetworkApiClient,
+    OverlayContentKey, StateContentKey, StateNetworkApiClient,
+};
+
+/// Ping delay for liveness check of peers in census
+/// Two minutes was chosen somewhat arbitrarily, and can be adjusted
+/// in the future based on performance
+const LIVENESS_CHECK_DELAY: Duration = Duration::from_secs(120);
+
+/// The maximum number of enrs to return in a response,
+/// limiting the number of OFFER requests spawned by the bridge
+/// for each piece of content
+const ENRS_RESPONSE_LIMIT: usize = 8;
+
+/// The census is responsible for maintaining a list of known peers in the network,
+/// checking their liveness, updating their data radius, iterating through their
+/// rfn to find new peers, and providing interested enrs for a given content key.
+pub struct Census {
```

Comment:
I have an idea, but feel free to ignore it if you don't like it.

Instead of having `history, state and beacon` network separately, we have one field: `networks` that is ether `HashMap<Subnetwork, Network>` of `Vec<Network>`.

The constructor would accept list of `Vec<Subnetwork>` and call `init()` for each of them (no need for separate `init()` function).

Reply:
Mmmm, personally I prefer to have the explicit subnetwork names... makes for easier access to the `Network`, rather than 'getting' from the hashmap, or implementing a getter fn. Though, I admit it's not a really big difference...